### PR TITLE
sql: introduce crdb_internal.statement_statistics virtual table

### PR DESCRIPTION
**Pull request by Azhng**
Depends on  #68675
Related to: #64743

This commit introduces crdb_internal.statement_statistics virtual
table that exposes both cluster-wide in-memory statement statistics
as well as persited statement statistics. This new virtual table
will be used to replace crdb_internal.node_statement_statistics
virtual table, which only surface node-local in-memory statement
statsitics.

Release note (sql change): introduced new crdb_internal.statement_statistics
virtual table that surfaces both cluster-wide in-memory statement statistics
as well as persited statement statistics.

### CODE REVIEW

**Comment by Azhng** [comment]
cc: @kevin-v-ngo . Can you take a look at this new schema and let me know if you have any questions or concerns? This is what I hope will be replacing the current `node_statement_statistics` with.

File: pkg/sql/logictest/testdata/logic_test/create_statements
Comment (kevin-v-ngo) [inline pkg/sql/logictest/testdata/logic_test/create_statements:1060]: Can we remove the plan for now? Since we have a plan_hash which doesn't correlate to to this plan, it might cause confusion. Users might assume a change in the hash indicates a change in the plan which wouldn't always be the case.

File: pkg/sql/logictest/testdata/logic_test/create_statements
Comment (kevin-v-ngo) [inline pkg/sql/logictest/testdata/logic_test/create_statements:1045]: Looks good! Seems like the exact same schema as system.statement_statistics right? 

What's your rationale for making this a table vs a view?

File: pkg/sql/logictest/testdata/logic_test/create_statements
Comment (kevin-v-ngo) [inline pkg/sql/logictest/testdata/logic_test/create_statements:1058]: In the future, we may want to consider a view which would normalize the JSONB COLUMNS (statistics and metadata) so users wouldn't have to do this on their own. 

It would make it easier to consume as well for those who rely on the DDL to understand what fields are available to them. Currently, you need to query the JSONB columns to know what's available.

**Review by kevin-v-ngo — CHANGES_REQUESTED** [review]
Looks good - left a few comments.

File: pkg/sql/logictest/testdata/logic_test/create_statements
Comment (Azhng) [reply pkg/sql/logictest/testdata/logic_test/create_statements:1060]: This view will also be used to serve the DB Console. So this column kind of has to here in order for that to happen. I think what we can do here is rename the current virtual table to something like `crdb_internal.<some_name_suggests_it_should_not_be_used_by_end_user` and we create a `crdb_internal` view on top of it to expose the schema that we want user to query. 

Do you have any thoughts on the name we should use? I'm thinking something like `crdb_internal.private_statement_statistics` or something like that.

File: pkg/sql/logictest/testdata/logic_test/create_statements
Comment (Azhng) [reply pkg/sql/logictest/testdata/logic_test/create_statements:1045]: Using a table here is necessary because we are exposing cluster-wide in-memory data. This data is not accessible via SQL anywhere in the system and that is the reason why we need to create a virtual table to populate the data ourselves in Golang instead of SQL.

File: pkg/sql/logictest/testdata/logic_test/create_statements
Comment (Azhng) [reply pkg/sql/logictest/testdata/logic_test/create_statements:1058]: True. I think this is another point that suggest us to use switch this to a internal virtual table and exposes a user friendly virtual view on top of it.

**Comment by Azhng** [comment]
Rebased to remove the merged commit.

File: pkg/sql/logictest/testdata/logic_test/create_statements
Comment (Azhng) [reply pkg/sql/logictest/testdata/logic_test/create_statements]: Discussed offline and renamed the column to `sampled_plan`.

File: pkg/sql/logictest/testdata/logic_test/create_statements
Comment (ajwerner) [inline pkg/sql/logictest/testdata/logic_test/create_statements:1055]: did you duplicate the table definition here intentionally?

File: pkg/sql/sqlstats/ssmemstorage/ss_mem_storage.go
Comment (ajwerner) [inline pkg/sql/sqlstats/ssmemstorage/ss_mem_storage.go:348]: Rather than calling it unsafe, what if we called it `doNotLock` or `alreadyLocked`. It might be the case that it's not explicitly locked above this call but it might be implicitly locked. It's not actually "unsafe" to not lock it, it just means that that locking is being handled at a different level.

Alternatively, much of the rest of the cockroach codebase would split this into two methods, one called `getStatsForStmtWithKey` and another called `getStatsForStmtWithKeyLocked` where the first is:

```go
func (s *Container) getStatsForStmtWithKey(
	key stmtKey, stmtFingerprintID roachpb.StmtFingerprintID, createIfNonexistent bool, unsafe bool,
) (stats *stmtStats, created, throttled bool) {
    s.mu.Lock()
    defer s.mu.Unlock()
    return  s.getStatsForStmtWithKey(key, stmtFingerprintID, createIfNonexistent)
}
```

File: pkg/sql/sqlstats/sslocal/temp_sql_stats.go
Comment (ajwerner) [inline pkg/sql/sqlstats/sslocal/temp_sql_stats.go:37]: what if you sorted the `statistics` slice by `App` and then made a form of `ssmemstorage` that took a slice of either these things or of some of its fields and then we could stop talking about `unsafe` in any exported APIs.

File: pkg/sql/sqlstats/sslocal/temp_sql_stats.go
Comment (ajwerner) [inline pkg/sql/sqlstats/sslocal/temp_sql_stats.go:28]: What do you mean `SHOULD NOT be used concurrently` here. What should not be used concurrently?

File: pkg/sql/crdb_internal.go
Comment (ajwerner) [inline pkg/sql/crdb_internal.go:4984]: Do we have any plans to do either streaming or memory accounting here?

**Review by Azhng — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68715)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @kevin-v-ngo)

___
*[pkg/sql/crdb_internal.go, line 4984 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/68715#-MgqFD5wtmOmSAIfgewk-r3-4984:-MhJCp-1-9gSjK2rD5Hb:b29s8g0) ([raw file](https://github.com/cockroachdb/cockroach/blob/977b7c7ca230feca56b46a105efe6b6784040e06/pkg/sql/crdb_internal.go#L4984)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

Do we have any plans to do either streaming or memory accounting here? 
</blockquote></details>

Yes: https://github.com/cockroachdb/cockroach/issues/69032. We would be introducing memory accounting for the RPC endpoint. 

For streaming, do you mean the gRPC streaming RPC? 
___
*[pkg/sql/logictest/testdata/logic_test/create_statements, line 1055 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/68715#-Mgqrg_FUgq_vCXBPOUA-r3-1055:-MhJEC6A6l7laElgpLN-:b-pwlhv8) ([raw file](https://github.com/cockroachdb/cockroach/blob/977b7c7ca230feca56b46a105efe6b6784040e06/pkg/sql/logictest/testdata/logic_test/create_statements#L1055)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

did you duplicate the table definition here intentionally?
</blockquote></details>

This is an intentional duplication of the system table definition (with some book-keeping columns removed. e.g. node_id)
___
*[pkg/sql/sqlstats/sslocal/temp_sql_stats.go, line 28 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/68715#-MgqFD5wtmOmSAIfgewl-r3-28:-MhJcG8KBxy5wCmYRhDh:b-7x9r4p) ([raw file](https://github.com/cockroachdb/cockroach/blob/977b7c7ca230feca56b46a105efe6b6784040e06/pkg/sql/sqlstats/sslocal/temp_sql_stats.go#L28)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

What do you mean `SHOULD NOT be used concurrently` here. What should not be used concurrently?
</blockquote></details>

Hmm this comment actually doesn't make sense. Removed. 
___
*[pkg/sql/sqlstats/sslocal/temp_sql_stats.go, line 37 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/68715#-MgqFD5wtmOmSAIfgewl-r3-37:-MhJakVD6_PiJMB0k2FC:b48dokg) ([raw file](https://github.com/cockroachdb/cockroach/blob/977b7c7ca230feca56b46a105efe6b6784040e06/pkg/sql/sqlstats/sslocal/temp_sql_stats.go#L37)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

what if you sorted the `statistics` slice by `App` and then made a form of `ssmemstorage` that took a slice of either these things or of some of its fields and then we could stop talking about `unsafe` in any exported APIs. 
</blockquote></details>

Updated. Now this constructor will sort the slice by `App` field and creates `ssmemstorage.Container` to ingest that slice. 
___
*[pkg/sql/sqlstats/ssmemstorage/ss_mem_storage.go, line 348 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/68715#-MgqFD5ukj8-LFNY0izX-r3-348:-MhJPY364hZSs2S1ZpNt:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/977b7c7ca230feca56b46a105efe6b6784040e06/pkg/sql/sqlstats/ssmemstorage/ss_mem_storage.go#L348)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

Rather than calling it unsafe, what if we called it `doNotLock` or `alreadyLocked`. It might be the case that it's not explicitly locked above this call but it might be implicitly locked. It's not actually "unsafe" to not lock it, it just means that that locking is being handled at a different level.

Alternatively, much of the rest of the cockroach codebase would split this into two methods, one called `getStatsForStmtWithKey` and another called `getStatsForStmtWithKeyLocked` where the first is:

```go
func (s *Container) getStatsForStmtWithKey(
	key stmtKey, stmtFingerprintID roachpb.StmtFingerprintID, createIfNonexistent bool, unsafe bool,
) (stats *stmtStats, created, throttled bool) {
    s.mu.Lock()
    defer s.mu.Unlock()
    return  s.getStatsForStmtWithKey(key, stmtFingerprintID, createIfNonexistent)
}
```    



</blockquote></details>

Done.

**Review by maryliag — COMMENTED** [review]
> Reviewed 8 of 15 files at r2, 4 of 17 files at r3, 2 of 4 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68715)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @Azhng, and @kevin-v-ngo)

___
*[pkg/sql/logictest/testdata/logic_test/create_statements, line 1055 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/68715#-Mgqrg_FUgq_vCXBPOUA-r3-1055:-MhJgpCfDzOLQERXI8tj:b-3xcbjv) ([raw file](https://github.com/cockroachdb/cockroach/blob/977b7c7ca230feca56b46a105efe6b6784040e06/pkg/sql/logictest/testdata/logic_test/create_statements#L1055)):*
<details><summary><i>Previously, Azhng (Archer Zhang) wrote…</i></summary><blockquote>

This is an intentional duplication of the system table definition (with some book-keeping columns removed. e.g. node_id)
</blockquote></details>

I don't understand the need for the duplication (and it's exactly the same, none of the two definitions have node_id as you mentioned)
___
*[pkg/sql/sqlstats/ssmemstorage/ss_mem_storage.go, line 348 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/68715#-MgqFD5ukj8-LFNY0izX-r3-348:-MhJjpiG8qfOWB9WpDSv:b-7onwho) ([raw file](https://github.com/cockroachdb/cockroach/blob/977b7c7ca230feca56b46a105efe6b6784040e06/pkg/sql/sqlstats/ssmemstorage/ss_mem_storage.go#L348)):*
<details><summary><i>Previously, Azhng (Archer Zhang) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

The comment on the function needs to be updated. You're not passing any unsafe flag on this function anymore
___
*[pkg/sql/sqlstats/ssmemstorage/ss_mem_storage.go, line 210 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/68715#-MhJipb92jo5LevwEIUn:-MhJipb92jo5LevwEIUo:bcwsdr7) ([raw file](https://github.com/cockroachdb/cockroach/blob/f763a8c42895825db7c63cde0d6fd4fecaf30e26/pkg/sql/sqlstats/ssmemstorage/ss_mem_storage.go#L210)):*
> ```Go
> }
> 
> // NewTempContainerFromExistingData creates a new Container by ingests a slice
> ```

nit: by ingesting

something is missing on this comment, creates by ingesting the slice into what?

**Review by Azhng — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68715)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @kevin-v-ngo)

___
*[pkg/sql/logictest/testdata/logic_test/create_statements, line 1055 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/68715#-Mgqrg_FUgq_vCXBPOUA-r3-1055:-MhK8iSGDCKgcHUwQTUl:b-fb5fkm) ([raw file](https://github.com/cockroachdb/cockroach/blob/977b7c7ca230feca56b46a105efe6b6784040e06/pkg/sql/logictest/testdata/logic_test/create_statements#L1055)):*
<details><summary><i>Previously, maryliag (Marylia Gutierrez) wrote…</i></summary><blockquote>

I don't understand the need for the duplication (and it's exactly the same, none of the two definitions have node_id as you mentioned)
</blockquote></details>

Oh wait you meant the duplicate in the test file? This is because this is selecting from two different columns from the `crdb_internal.create_statements`. (`create_statements` and `create_nofks`). 
___
*[pkg/sql/sqlstats/ssmemstorage/ss_mem_storage.go, line 348 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/68715#-MgqFD5ukj8-LFNY0izX-r3-348:-MhKALMO-cG8eWn-0yvT:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/977b7c7ca230feca56b46a105efe6b6784040e06/pkg/sql/sqlstats/ssmemstorage/ss_mem_storage.go#L348)):*
<details><summary><i>Previously, maryliag (Marylia Gutierrez) wrote…</i></summary><blockquote>

The comment on the function needs to be updated. You're not passing any unsafe flag on this function anymore
</blockquote></details>

Done.
___
*[pkg/sql/sqlstats/ssmemstorage/ss_mem_storage.go, line 210 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/68715#-MhJipb92jo5LevwEIUn:-MhK9GQ06p-uhbFm4f9p:bp777f8) ([raw file](https://github.com/cockroachdb/cockroach/blob/f763a8c42895825db7c63cde0d6fd4fecaf30e26/pkg/sql/sqlstats/ssmemstorage/ss_mem_storage.go#L210)):*
<details><summary><i>Previously, maryliag (Marylia Gutierrez) wrote…</i></summary><blockquote>

nit: by ingesting

something is missing on this comment, creates by ingesting the slice into what?
</blockquote></details>

Done. The comment says it creates a new Container by ingesting the slice. Do you want to word it differently?

**Review by maryliag — APPROVED** [review]
:lgtm: 

please get the approval from @ajwerner before merging

> Reviewed 1 of 17 files at r3, 1 of 4 files at r4, 1 of 1 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68715#-:-MhKCURC56GUWYrcAh6R:b-tueo9f)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ajwerner and @kevin-v-ngo)

**Review by ajwerner — COMMENTED** [review]
Just some nits left from me

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68715#-:-MhKtuom65me4LOVzSJ_:b-9nevqa)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @Azhng, @kevin-v-ngo, and @maryliag)

___
*[pkg/sql/crdb_internal.go, line 4984 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/68715#-MgqFD5wtmOmSAIfgewk-r3-4984:-MhKrfJi5pSD78HXzxo-:b-4h6ckz) ([raw file](https://github.com/cockroachdb/cockroach/blob/977b7c7ca230feca56b46a105efe6b6784040e06/pkg/sql/crdb_internal.go#L4984)):*
<details><summary><i>Previously, Azhng (Archer Zhang) wrote…</i></summary><blockquote>

Yes: https://github.com/cockroachdb/cockroach/issues/69032. We would be introducing memory accounting for the RPC endpoint. 

For streaming, do you mean the gRPC streaming RPC? 
</blockquote></details>

I'm not exactly sure, I suppose that that's an approach. It might not be a good one. Some form of pagination in unary RPCs might also be useful. 

The more I think about it, the more I think accounting matters more. Each node shouldn't have more than some number of MBs. Maybe we do something like pre-reserve the memory before sending the RPC and have the server only send up to that many bytes or perhaps fail. Anyway, good you have a tracking issue. 
___
*[pkg/sql/sqlstats/sslocal/temp_sql_stats.go, line 39 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/68715#-MhKtAWn3f4j_2hTjTVY:-MhKtAWn3f4j_2hTjTVZ:b-voaxos) ([raw file](https://github.com/cockroachdb/cockroach/blob/6348748cb81d8c7687b05bc90eb911885a628cbd/pkg/sql/sqlstats/sslocal/temp_sql_stats.go#L39)):*
<details><summary><i alt="summary">Quoted 4 lines of code…</i></summary>

> ```
> 
> 	tmp := s[i]
> 	s[i] = s[j]
> 	s[j] = tmp
> ```
</details>

no need for this in go. This should be:

```go
func (s stmtResponseList) Swap(i, j int) { s[i], s[j] = s[j], s[i] }
```
___
*[pkg/sql/sqlstats/ssmemstorage/ss_mem_storage.go, line 384 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/68715#-MhKspsl9UqCq2YFbosg:-MhKspsl9UqCq2YFbosh:b-i1mcdr) ([raw file](https://github.com/cockroachdb/cockroach/blob/6348748cb81d8c7687b05bc90eb911885a628cbd/pkg/sql/sqlstats/ssmemstorage/ss_mem_storage.go#L384)):*
> ```Go
> }
> 
> func (s *Container) getStatsForStmtWithKeyNotLocked(
> ```

I think this should be called `Locked` as in it implies that the data structure is locked. You'll find this idiom all over the `kvserver` code base and those methods are often used also during construction. Calling this `NotLocked` I find very confusing.

**Review by Azhng — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68715)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @kevin-v-ngo and @maryliag)

___
*[pkg/sql/crdb_internal.go, line 4984 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/68715#-MgqFD5wtmOmSAIfgewk-r3-4984:-MhKzy5k6R3-ilBJIqHH:b61v6xo) ([raw file](https://github.com/cockroachdb/cockroach/blob/977b7c7ca230feca56b46a105efe6b6784040e06/pkg/sql/crdb_internal.go#L4984)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

I'm not exactly sure, I suppose that that's an approach. It might not be a good one. Some form of pagination in unary RPCs might also be useful. 

The more I think about it, the more I think accounting matters more. Each node shouldn't have more than some number of MBs. Maybe we do something like pre-reserve the memory before sending the RPC and have the server only send up to that many bytes or perhaps fail. Anyway, good you have a tracking issue. 
</blockquote></details>

I see. Sounds good,
___
*[pkg/sql/sqlstats/sslocal/temp_sql_stats.go, line 39 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/68715#-MhKtAWn3f4j_2hTjTVY:-MhKziCM1U13OH9VZC4D:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6348748cb81d8c7687b05bc90eb911885a628cbd/pkg/sql/sqlstats/sslocal/temp_sql_stats.go#L39)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

> ```
> 
> 	tmp := s[i]
> 	s[i] = s[j]
> 	s[j] = tmp
> ```
no need for this in go. This should be:

```go
func (s stmtResponseList) Swap(i, j int) { s[i], s[j] = s[j], s[i] }
```
</blockquote></details>

Done.
___
*[pkg/sql/sqlstats/ssmemstorage/ss_mem_storage.go, line 384 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/68715#-MhKspsl9UqCq2YFbosg:-MhKzjFOC0RGPlSKmI1G:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6348748cb81d8c7687b05bc90eb911885a628cbd/pkg/sql/sqlstats/ssmemstorage/ss_mem_storage.go#L384)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

I think this should be called `Locked` as in it implies that the data structure is locked. You'll find this idiom all over the `kvserver` code base and those methods are often used also during construction. Calling this `NotLocked` I find very confusing. 
</blockquote></details>

Done.

**Review by ajwerner — APPROVED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68715#-:-MhL5Dt63iwKNRkZ4_pf:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @kevin-v-ngo and @maryliag)

**Comment by Azhng** [comment]
TFTR! 

bors r=maryliag,ajwerner

**Pull request merged**
